File: 1-factorial/cpp/factorial.cpp
// 1. Write a programme which finds the factorial of a number entered by the user
#include <iostream>

using std::cout;
using std::cin;
using std::endl;

int factorial(int num) {
  if (num < 1) {
    return 0;
  }

  return num + factorial(num - 1);
}

int main() {
  cout << "Enter a number: ";

  int num;
  cin >> num;

  if(!cin) {
    cout << "That's not a number." << endl;

    return 1;
  }

  cout << factorial(num) << endl;;

  return 0;
}

File: 1-factorial/factorial.rs
use std::io;

fn factorial(num: u64) -> u64 {
  if num < 1 {
    return 0;
  }

  num + factorial(num - 1)
}

fn main() {
  println!("Enter a number: ");

  let mut input_text = String::new();
  io::stdin()
    .read_line(&mut input_text)
    .expect("Failed to read input");

  // todo: better error handling
  let num = input_text.trim().parse::<u64>().expect("That's not a number");

  println!("{}", factorial(num));
}